specification: updating for latest linux changes (#35)
TrenchBoot/documentation: specifications/secure-launch-specification.rst

Secure Launch Interfaces
========================

Secure Launch Interfaces consist of three interfaces, the DLME Header, the DLE
Handler Specifications, and the SLRT Specification.

DLME Header
-----------

Detailed below is the header structure that a Secure Launch-compatible kernel
must provide and is discoverable. The information provides details to the DLE
Handler to facilitate it in launching the kernel.

> DLME Header Structure

Reviewer: Are there any alignment requirements?

Author: The only requirement from TXT spec is that it must be within the MLE range. If we want to impose alignment, I'm open to that discussion. It may require a slight change to the Linux series, though I don't think it would be difficult.

Reviewer: I'm just thinking of AArch64 and other platforms that don't allow misaligned accesses.  If you don't think this will cause problems when storing/reading the structure, then can explicitly say that the structure is 1-byte aligned.

~~~~~~~~~~~~~~~~~~~~~

A descripton of the DLME for the DLE Handler.

:uuid0: Chunk 0 of UUID, 9082AC5A
:uuid1: Chunk 1 of UUID, 74A7476F
:uuid2: Chunk 2 of UUID, A2555C0F

> :uuid3: Chunk 3 of UUID, 42B651CB

Reviewer 2: Should it be unconditionally included in non-TXT binaries as well? It probably makes sense for x86 as vendor can be easily detected, but I'm not so sure about other architectures.

Author: Yes, while it originates from Intel TXT, we want to standardize the dl-stub across architectures and this provides the cleanest, IOW maximize reuse, method to communicate the necessary information to the dl-stub

> :header_len: Length of the header.

Reviewer: To be unambiguous since this is not always the case:
```suggestion
:header_len: Length of the header in bytes.
```

Author: ack

:version: Header version.
:entry_point: Address of the DLME entry point.

> :first_valid_page: Address to first valid page of DLME.

Reviewer: ```suggestion
:first_valid_page: Address of the first valid 4KiB-page of DLME.
```
I suppose it also needs to have zeroes in its lowest bits.

Author: I would not specify 4K, it should be kept platform agnostic. I am good with the other wording change.

Reviewer: Then could say `page of a platform-specific size` so a reader doesn't look for the page size specified somewhere else in the specification.

:start: Offset to first byte of DLME within page space.

> :end: Offset to last byte + 1 of DLME within page space.

Reviewer: The base for the offsets is unspecified.

Author: This is how it is defined in MLE guide,

_Note: linear addresses start at the beginning of the executable region of the binary,
not the beginning of the file. The offset between the beginning of the binary file and
beginning of the linear address space is ABI-dependent._

I am going to have to think of a good way to say this in relation to this spec. Suggestions are welcome.

Platform Requirements
~~~~~~~~~~~~~~~~~~~~~

| **1** - General Requirements
| **1.1** - The DLME **SHALL** provide a DLME Header
| **1.2** - The DLME **SHALL** provide a mechanism to discover the DLME Header
|
| **2** - x86 Non-EFI Platforms
| **2.1** - The DLE Handler **MAY** be invoked with the CPU in either 32bit
|       protected mode or 64bit long mode
| **2.2** - The SLRT **SHALL** be passed to the DLE Handler in the EDI/RDI CPU
|       register

> | **2.3** - All other registers besides EDI/RDI are not guarenteed

Reviewer: A typo:
```suggestion
| **2.3** - All other registers besides EDI/RDI are not guaranteed
```

Author: ack

| **2.4** - The invoking code **SHALL** use a long jump to the DLE Handler
| **2.5** - The DLE Handler **SHALL NOT** return control on error
|
| **3** - x86 EFI Platforms
| **3.1** - The DLE Handler **SHALL** implement the Secure Launch EFI Protocol.
|
| **2** - Arm Platforms
| **2.1** - *Reserved*

EFI SLaunch Protocol
~~~~~~~~~~~~~~~~~~~~

On EFI platforms the DLE Handler must implement the EFI_SLAUNCH_PROTOCOL that provides the following methods.

* **EFI_SLAUNCH_REGISTER_DLME**: Registers the DLME with the DLE Handler
* **EFI_SLAUNCH_LAUNCH**: Invokes the DLE Handler

The interface definition for the protocol is as follows,

.. code-block:: c
    :linenos: 1

    typedef struct tdEFI_SLAUNCH_PROTOCOL {
      EFI_SLAUNCH_SLRT_UPDATE_REGISTER_DLME RegisterDlme;
      EFI_SLAUNCH_LAUNCH Launch;

>     } EFI_SLAUNCH_PROTOCOL

Reviewer: ```suggestion
    } EFI_SLAUNCH_PROTOCOL;
```

Author: ack

    typedef EFI_STATUS (EFIAPI *EFI_SLAUNCH_REGISTER_DLME) (
                        IN EFI_SLAUNCH_PROTOCOL *This,

>                         IN EFI_PHYSICAL_ADDRESS *DlmeBase
>                         IN EFI_PHYSICAL_ADDRESS *DlmeHeader

Reviewer: ```suggestion
                        IN EFI_PHYSICAL_ADDRESS *DlmeBase,
                        IN EFI_PHYSICAL_ADDRESS *DlmeHeader,
```

Author: ack

:dlme_size: The size of the DLME.
:dlme_base: The base address where the DLME is located.
:dlme_entry: The offset into the DLME of the entry point.

> :dlme_header_offset: The offest into the DLME of the DLME Header.
> :dlme_config_table: The offsee into the DLME of the DLME Config Table.

Reviewer: ```suggestion
:dlme_header_offset: The offset into the DLME of the DLME Header.
:dlme_config_table: The offset into the DLME of the DLME Config Table.
```

Author: ack

    struct os2mle {
        u32 version;
        u32 reserved;
        u64 slrt;

>         u64 txt_info; 

Reviewer 2: Remove trailing space.

Author: ack.
